source/agora/utils/InetUtils.d

```diff
@@ -0,0 +1,224 @@
+            uint ifa_flags;
+            /// Address of interface
+            sockaddr* ifa_addr;
+            /// Netmask of interface
+            sockaddr* ifa_netmask;
+            /// Point-to-point destination addresss
+            sockaddr* if_dstaddr;
+            /// Address specific data
+            void* ifa_data;
+        }
+
+        /// Returns: linked list of ifaddrs structures describing interfaces
+        int getifaddrs(ifaddrs** );
+        /// Frees the linked list returned by getifaddrs
+        void freeifaddrs(ifaddrs* );
+    }
+    else
+    {
+        import core.sys.darwin.ifaddrs;
+    }
+}
+
+struct InetUtils
+{
+version(OSX)
+{
+    import core.sys.posix.netdb;
+    import core.sys.posix.netinet.in_;
+    import core.sys.posix.sys.socket;
+
+    mixin AddGetAllIPs;
+}
+version(linux)
+{
+    import core.sys.linux.ifaddrs;
+    import core.sys.posix.netdb;
+
+    mixin AddGetAllIPs;
+}
+version (Windows)
+{
+    import std.socket;
+
+    import core.sys.windows.iphlpapi;
+    import core.sys.windows.iptypes;
+    import core.sys.windows.windef;
+    import core.sys.windows.winsock2;
+    import core.stdc.stdlib: malloc, free;
+    import core.stdc.string: strlen;
+
+    public static string[] getAllIPs()
+    {
+        string[] ips;
+        PIP_ADAPTER_INFO adapter_info_head;
+        PIP_ADAPTER_INFO adapter_info;
+        DWORD ret_adapters_info;
+        ULONG buff_length = IP_ADAPTER_INFO.sizeof;
+        adapter_info_head = cast(IP_ADAPTER_INFO *) malloc(IP_ADAPTER_INFO.sizeof);
+        if (adapter_info_head == NULL)
+        {
+            log.error("Error allocating memory needed to call GetAdaptersinfo - 1");
+            return [];
+        }
+        scope(exit) free(adapter_info_head);
+        // find out the real size we need to allocate
+        if (GetAdaptersInfo(adapter_info_head, &buff_length) == ERROR_BUFFER_OVERFLOW)
+        {
+            free(adapter_info_head);
+            adapter_info_head = cast(IP_ADAPTER_INFO *) malloc(buff_length);
+            if (adapter_info_head == NULL)
+            {
+                log.error("Error allocating memory needed to call GetAdaptersinfo - 2");
+                return [];
+            }
+        }
+        if ((ret_adapters_info = GetAdaptersInfo(adapter_info_head, &buff_length)) == NO_ERROR) {
+            adapter_info = adapter_info_head;
+            while (adapter_info)
+            {
+                auto ip_tmp = cast(char *) adapter_info.IpAddressList.IpAddress.String;
+                string ip = ip_tmp[0 .. strlen(ip_tmp)].idup;
+                if(!ip.length || ip == "0.0.0.0")
+                {
+                    adapter_info = adapter_info.Next;
+                    continue;
+                }
+                ips ~= ip;
+                adapter_info = adapter_info.Next;
+            }
+        }
+        else
+        {
+            log.error("GetAdaptersInfo failed with error: {}", ret_adapters_info);
```

Comment:
```suggestion
            log.error("GetAdaptersInfo failed with error: {}", ret_adapters_info);
            return null;
```

And the `return ips;` below should be moved up into the last line of the `if ((ret_adapters_info = GetAdaptersInfo(adapter_info_head, &buff_length)) == NO_ERROR)` body.

Reply:
I think it is good if we only have one exit point from a function which is usually at the end of the function. Sometimes like when there is an exceptional case we can return early.

Is there any rationale behind returning in the if statement?

Reply:
If you have only one exit point and it returns a variable that is declared at the top of the function then you don't know what you're returning without inspecting the code flow.

Reply:
> If you have only one exit point and it returns a variable that is declared at the top of the function then you don't know what you're returning without inspecting the code flow.

that is true, but lets say someone later refactors the code and add some filtering or manipulation code to the 'ips' variable after the premature return. In this case he will need to change premature return in every line which is error phrone.

Anyways now I see there are pros and cons, I will follow your suggestion and return early

Reply:
I see your point. In this case I would argue that:

- We should handle all error conditions first.
- If the function becomes so big that it's easy to miss there was some filtering done afterwards then the function should be split into multiple smaller functions. We do not want 1000 line functions in our code.